src/UniDeposit.sol

```diff
@@ -53,64 +55,68 @@ contract UniDeposit {
         _;
     }
 
+    function unminted_USDT() view public returns (uint) {
+        return IERC20(USDT).balanceOf(address(this));
+    }    
+    function minted_yCRV() view public returns (uint) {
+        return IERC20(yCrvToken).balanceOf(address(this));
+    }
     function get_yCrvFromUsdt(uint256 amount) public view returns (uint256) {
-        return amount.mul(minted_yCRV).div(mintedUSDT);
+        return amount.mul(minted_yCRV()).div(mintedUSDT);
     }
+    function get_usdtFromYcrv(uint256 amount) public view returns (uint256) {
+        return amount.mul(mintedUSDT).div(minted_yCRV());
+    }    
 
     event Deposit(address indexed who, uint256 amountOfUsdt);
 
-    function deposit(uint256 amount) public goodToChargeUSDT(amount) {
+    function deposit(uint256 input) external goodToChargeUSDT(input) {
         // sadly no return from USDT
-        IUSDT(USDT).transferFrom(msg.sender, address(this), amount);
-        if (amount > mintedUSDT) {
+        IUSDT(USDT).transferFrom(msg.sender, address(this), input);
+        if (input > mintedUSDT) {
             // New Deposit goes to unmintedUSDT pool
-            unmintedUSDT = unmintedUSDT.add(amount);
-            emit Deposit(msg.sender, amount);
-            setBalance(msg.sender, balanceOf(msg.sender).add(amount));
+            setBalance(msg.sender, balanceOf(msg.sender).add(input));
+            emit Deposit(msg.sender, input);
         } else {
             // if enough, just swap then
-            uint256 yCrvWillGet = get_yCrvFromUsdt(amount);
-            IERC20(yCrvToken).transfer(msg.sender, yCrvWillGet);
+            uint256 output = get_yCrvFromUsdt(input);
+            mintedUSDT = mintedUSDT.sub(input);            
+            IERC20(yCrvToken).transfer(msg.sender, output);
         }
     }
 
+    function withdraw(uint input) external {
+        uint ycrv = minted_yCRV();
+        require(input <= ycrv, "Insufficient minted yCrv.");
+        uint output = get_usdtFromYcrv(input);
+        mintedUSDT = mintedUSDT.sub(output);
+        IERC20(yCrvToken).transferFrom(msg.sender, address(this), input);
+        IUSDT(USDT).transfer(msg.sender, output);
+    }
+
+    // The world could always use more heroes.
     function mint() public {
-        uint256 yCrvBalanceBeforeMint = IERC20(yCrvToken).balanceOf(address(this));
-        // Minting now
-        // I don't know how to set this val
-        uint256 min_mint_amount = 0;
-
-        IUSDT(USDT).approve(yDeposit, unmintedUSDT);
```

Comment:
You *should check* USDT allowance for `yDeposit` to spend USDT on behalf of `UniDeposit`, before addLiquidity to `yDeposit`. 
Even though you [approve `UINT_MAX` in the constructor](https://github.com/frankwei98/My-solidity-workshop/pull/1/files#diff-0afddb572d15ce58ef43d808d092aad3R40) already, there are chances that allowance is depleted.

Reply:
I think this function needs to fix

Reply:
🏳️